### ISSUE TITLE
Unused variable warning for used variable.

### ISSUE THREAD

**Issue by koehlma** [rust-lang/rust#85071]
When trying to compile the following code I get an unused variable warning despite the variable is being used:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=66167b96f0e14d53616abbd0e6b3d12c

The current output is:
```
warning: unused variable: `typ`
  --> src/lib.rs:62:17
   |
62 |             let typ = T::infer_type(self);
   |                 ^^^ help: if this is intentional, prefix it with an underscore: `_typ`
   |
   = note: `#[warn(unused_variables)]` on by default
```

Ideally the output should not produce such a warning.

`cargo fix` produces the following output:

```
warning: failed to automatically apply fixes suggested by rustc to crate `rust_playground`

after fixes were automatically applied the compiler reported errors within these files:

  * src\lib.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `typ` in this scope
  --> src\lib.rs:64:30
   |
64 |             info.typ.replace(typ);
   |                              ^^^ help: a local variable with a similar name exists: `_typ`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: unused variable: `typ`
  --> src\lib.rs:62:17
   |
62 |             let typ = T::infer_type(self);
   |                 ^^^ help: if this is intentional, prefix it with an underscore: `_typ`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted

warning: unused variable: `typ`
  --> src\lib.rs:62:17
   |
62 |             let typ = T::infer_type(self);
   |                 ^^^ help: if this is intentional, prefix it with an underscore: `_typ`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.04s
```

**Comment by ehuss** [comment]
Yea, since an empty enum cannot be created, I presume the compiler assumes all of the code after the call to `infer_type` cannot be run, so it does not exist.  So, in a way, `typ` cannot be used because it cannot be constructed.  I kinda wonder why it didn't give an unreachable statement warning instead.  Anyways, yea, the warning could be better.

**Comment by FabianWolff** [comment]
Yes, this behavior seems to come from this part of the liveness analysis:
https://github.com/rust-lang/rust/blob/ff34b919075f35a1787659e9c448a34b06bab8de/compiler/rustc_passes/src/liveness.rs#L979-L1007
i.e. the compiler realizes that a call to a method returning an uninhabited type can never return (it has to either panic or be non-terminating), and so it discards everything after the call for the purposes of liveness analysis. I agree that the warnings about unused variables, while technically correct, are somewhat confusing, and that there should probably be a warning about unreachable statements after the call.

**Comment by koehlma** [comment]
Makes sense. Yes, an unreachable statement warning for the remainder of the code would be more helpful. 👌

**Comment by hi-rustin** [comment]
@rustbot claim

Let me try to add a warning.

**Comment by FabianWolff** [comment]
> Let me try to add a warning.

Good luck, though I tried to look into this yesterday and concluded that there is no easy way to do this. Checking for unreachable expressions is integrated into the type checker, as divergence can influence type inference, e.g. here:
https://github.com/rust-lang/rust/blob/bba8710616e5e4722215c0d6b27abaedca03ebad/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs#L654-L660
Unfortunately, this means that full type information is not yet available at the point where the warnings for unreachable statements are generated. The easy solution would have been to add a call to `tcx.is_ty_uninhabited_from()` here:
https://github.com/rust-lang/rust/blob/bba8710616e5e4722215c0d6b27abaedca03ebad/compiler/rustc_typeck/src/check/expr.rs#L202-L205
Predictably, though, this doesn't work, because type information hasn't been stored in `tcx` yet (see [here](https://github.com/rust-lang/rust/blob/bba8710616e5e4722215c0d6b27abaedca03ebad/compiler/rustc_typeck/src/check/mod.rs#L37-L49)). It doesn't even crash, but leads to a broken (non-bootstrappable) compiler.

One possibility would be to move warnings about unreachable code into a separate pass, but then there would be code duplication between the type checker's divergence checking and this separate pass; maybe that is worth it if there is some other benefit, but functions returning uninhabited types seems like too much of an edge case to be worth it.

I don't mean to discourage you, though — maybe you can find a better solution!

**Comment by hi-rustin** [comment]
I can't fix it. So unassign myself.
